File: linked-list-cycle-ii/linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        if not head: return None
        seen = []
        while head:
            if head not in seen: 
                seen.append(head)
                head = head.next
            else:
                return head
        return None

File: linked-list-cycle-ii/linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        auto fast = head;
        auto slow = head;
        
        while (fast){
            if (!fast->next) return nullptr;
            fast = fast->next->next;
            slow = slow->next;
            if (fast == slow) break;
        }
        if (!fast) return nullptr;
        auto tmp = head;
        while (true){
            if (slow == tmp) return slow;
            slow = slow->next;
            tmp = tmp->next;
        }
    }
};
